sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/throughputControl/controller/container/ThroughputContainerController.java

```diff
@@ -0,0 +1,321 @@
+import com.azure.cosmos.implementation.changefeed.CancellationTokenSource;
+import com.azure.cosmos.implementation.throughputControl.ThroughputResolveLevel;
+import com.azure.cosmos.implementation.throughputControl.controller.group.ThroughputGroupControllerBase;
+import com.azure.cosmos.implementation.throughputControl.controller.group.ThroughputGroupControllerFactory;
+import com.azure.cosmos.models.CosmosQueryRequestOptions;
+import com.azure.cosmos.models.ModelBridgeInternal;
+import com.azure.cosmos.models.ThroughputProperties;
+import com.azure.cosmos.models.ThroughputResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+import reactor.core.scheduler.Schedulers;
+import reactor.util.retry.RetrySpec;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkArgument;
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
+
+/**
+ * Three main purpose of container controller:
+ * 1. Resolve max container throughput
+ * 2. Create and initialize throughput group controllers
+ * 3. Start throughput refresh task if necessary
+ */
+public class ThroughputContainerController implements IThroughputContainerController {
+    private static final Logger logger = LoggerFactory.getLogger(ThroughputContainerController.class);
+
+    private static final Duration DEFAULT_THROUGHPUT_REFRESH_INTERVAL = Duration.ofMinutes(15);
+    private static final int NO_OFFER_EXCEPTION_STATUS_CODE = HttpConstants.StatusCodes.BADREQUEST;
+    private static final int NO_OFFER_EXCEPTION_SUB_STATUS_CODE = HttpConstants.SubStatusCodes.UNKNOWN;
+
+    private final AsyncDocumentClient client;
+    private final ConnectionMode connectionMode;
+    private final GlobalEndpointManager globalEndpointManager;
+    private final ConcurrentHashMap<String, ThroughputGroupControllerBase> groupControllers;
+    private final List<ThroughputControlGroup> groups;
+    private final AtomicReference<Integer> maxContainerThroughput;
+    private final RxPartitionKeyRangeCache partitionKeyRangeCache;
+    private final CosmosAsyncContainer targetContainer;
+
+    private final CancellationTokenSource cancellationTokenSource;
+    private final Scheduler scheduler;
+
+    private ThroughputGroupControllerBase defaultGroupController;
+    private String targetContainerRid;
+    private String targetDatabaseRid;
+    private ThroughputResolveLevel throughputResolveLevel;
+
+    public ThroughputContainerController(
+        ConnectionMode connectionMode,
+        GlobalEndpointManager globalEndpointManager,
+        List<ThroughputControlGroup> groups,
```

Comment:
This takes a `List` -> I thought implementation layer was working on a `Set` ?

Reply:
As discussed offline, changed the code to use set.